cmd/snap-confine/mount-support.c

```diff
@@ -357,10 +358,16 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 		// assumes to find snap-exec
 		sc_must_snprintf(dst, sizeof dst, "%s/usr/lib/snapd",
 				 scratch_dir);
-		// FIXME: take re-exec into account, i.e. this may not
-		//        actually be the path we need, we need snapd
-		//        to tell us what path to use instead.
-		const char *src = LIBEXECDIR;
+
+		// bind mount the current $ROOT/usr/lib/snapd path,
+		// where $ROOT is either "/" or the "/snap/core/current"
+		// that we are re-execing from
+		char *src = LIBEXECDIR;
+		char self[PATH_MAX + 1] = { 0, };
+		if (readlink("/proc/self/exe", self, sizeof self) > 0) {
+			src = dirname(self);
+		}
```

Comment:
I think we should `} else { die(...) }`. no? If the kernel is failing to give what we need, we probably don't want to continue using whatever LIBEXECDIR is set to, but maybe I am missing something? Not to mention, we are initializing src to LIBEXECDIR, but then we immediately reassign it to dirname so, barring a kernel bug, we are using extra memory on the stack for no reason.

Reply:
I was initially thinking that LIBEXECDIR is a valid fallback if for whatever reason we cannot read /proc - however if you think we should fail hard in this case I'm happy to add code for this.

Reply:
I updated the code to die() if /proc/self/exe cannot be read.

Reply:
I think it would make the most sense to use `char *src = NULL;` and require readlink() to succeed and when it does, `src = dirname(self)`.